### ISSUE TITLE
Shared variables with nested lambdas returning incorrect values

### ISSUE THREAD

**Issue by MattEdwardsWaggleBee** [dadhi/FastExpressionCompiler#437]
I found an issue with variables that are shared across lambdas.  

In the test below I would assume `myVar` would be `3` after it's assigned inside the lambda, but instead this returns `5`.  Not sure if this should be throwing `NotSupportedExpressionException` similar to: `ParameterIsNotVariableNorInPassedParameters`

```
[Test]
public void Nested_lambda_with_shared_variable()
{
    var myVar = Variable( typeof( int ), "myVar" );
    Expression<Action<Action>> invokeParamLambda = ( lambda ) => lambda();

    var withNestedLambda = Lambda<Func<int>>( 
        Block( 
            [myVar],
            Assign( myVar, Constant( 5 ) ),
            Invoke( invokeParamLambda, 
                Lambda<Action>( Assign( myVar, Constant( 3 ) ) ) 
            ),
            myVar
        )
    );
    var compileLambda = withNestedLambda.Compile();
    var result = compileLambda();  // returns 3

    var fastCompileLambda = withNestedLambda.CompileFast( false, CompilerFlags.ThrowOnNotSupportedExpression );
    var fastResult = fastCompileLambda();  // returns 5
        
    Assert.AreEqual( result, fastResult );
}
```

Thanks for the great library!

**Comment by dadhi** [comment]
Hi, thanks for the test. I will check.

**Comment by dadhi** [comment]
@MattEdwardsWaggleBee 

Yes, it is an old problem, but it did not have a dedicated issue. Now it has.

You can use the workaround by wrapping the changing value in a box, check the example in the test in my commit for this issue.
Not convenient, but nevertheless. 

Also, I have found an additional issue when tried to use `struct Box` where System Compile works, but FEC produces the invalid IL. Documenting it here to fix.

**Comment by MattEdwardsWaggleBee** [comment]
Thank you for your quick reply and the workaround.

**Comment by dadhi** [comment]
@MattEdwardsWaggleBee This thingy requires more thoughtful work to do. I will work on it after v5.0.1 release.

**Comment by MattEdwardsWaggleBee** [comment]
@dadhi, Do you want to move this to a new milestone? If so, I can retest with 5.0.1, and maybe find a workaround for this in my project.

**Comment by dadhi** [comment]
@MattEdwardsWaggleBee I am working on it right now. I think if not fixed Today, I will release the 5.0.1 Tomorrow.
